File: 125 Valid Palindrome (No Java yet)/validPalindrome.py
class Solution(object):        
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s) == 0 or len(s) == 1:
            return True
        else:
            l = 0
            r = len(s) - 1
            
            while l < r:
                while l < r and (not s[l].isalnum()):
                    l += 1
                while r > l and (not s[r].isalnum()):
                    r -= 1
                if s[l].lower() != s[r].lower():
                    return False
                elif s[l].lower() == s[r].lower():
                    l += 1
                    r -= 1
            
            if l >= r:
                   return True

# The way of defining functions in a class is totally different from Java or C++
# And python has built-in functions to determine if characters are alphanumeric

File: 125 Valid Palindrome (No Java yet)/validPalindrome.cpp
class Solution {
public:
    // Determine whether current element is alphanumeric
    bool isAlphanumeric(const char c){
        if(c >= '0' && c <= '9')
            return true;
        else if(c >= 'a' && c <= 'z')
            return true;
        else if(c >= 'A' && c <= 'Z')
            return true;
        else
            return false;
    } 
    // Convert upper case to lower case
    char toLower(const char c){
        if(c >= 'A' && c <= 'Z')
            return c+32;
        else
            return c;
    }
    bool isPalindrome(string s) {
        if(s.size() == 0 || s.size() == 1) //If the string is not longer than 2
            return true;
        
        else{ //When the string is longer than 2
            int l = 0;
            int r = s.size() - 1;
            
            while(l < r){
                while(l < r && (!isAlphanumeric(s[l])))
                    l++;
                while(r > l && (!isAlphanumeric(s[r])))
                    r--;
                if(toLower(s[l]) != toLower(s[r]))
                    return false;
                else if(toLower(s[l]) == toLower(s[r])){
                    l++;
                    r--;
                }
            }
            if(l >= r)
                return true;
        }
    }
};